docs/design/endpoints.md

```diff
@@ -0,0 +1,292 @@
+used in the ruleset is an example of a standard-library function used by endpoint resolvers.
+
+This particular ruleset is kept simple for the purpose of example, naturally, a ruleset can be of arbitrary complexity as
+required by the service.
+
+```json
+{
+  "parameters": {
+    "ResourceId": {
+      "type": "string",
+      "required": "true"
+    }
+  },
+  "rules": [
+    {
+      "documentation": "use a special endpoint for government resources",
+      "type": "endpoint",
+      "conditions": [
+        {
+          "fn": "substring",
+          "argv": [
+            {"ref": "ResourceId"},
+            0,
+            5,
+            false
+          ],
+          "assign": "resourceIdPrefix"
+        },
+        {
+          "fn": "stringEquals",
+          "argv": [
+            {"ref": "resourceIdPrefix"},
+            "gov."
+          ]
+        }
+      ],
+      "endpoint": {
+        "url": "https://gov.api"
+      }
+    },
+    {
+      "documentation": "fallback to global endpoint",
+      "conditions": [],
+      "endpoint": {
+        "url": "https://global.api"
+      }
+    }
+  ]
+}
+```
+*Note*: The signature of the substring method is `substring(s: String, start: Int, end: Int, reverseIndices: false): String`.
+The argv property in the corresponding function call describes how to source the arguments (or input constants directly
+in the case of the indices).
+
+Given a sample operation `GetResource`, with a single operation input `resource` bound to the `ResourceId` provider
+parameter via context traits, this ruleset would roughly generate the following code:
+
+```kotlin
+public data class SmithyServiceEndpointParams(
+    public val resourceId: String?,
+) {
+    init {
+        require(resourceId != null) { "param resourceId is required" }
+    }
+}
+
+public fun interface SmithyServiceEndpointProvider {
+    open suspend fun resolveEndpoint(params: SmithyServiceEndpointParams): Endpoint
+}
+
+public class DefaultSmithyServiceEndpointProvider: SmithyServiceEndpointProvider {
+    public suspend fun resolveEndpoint(params: SmithyServiceEndpointParams): Endpoint {
+        val resourceIdPrefix = substring(params.resourceId, 0, 5, false)
+        if (resourceIdPrefix == "gov.") {
+            // use a special endpoint for government resources
+            return Endpoint(
+                Url.parse("https://gov.api"),
+            )
+        }
+
+        // fallback to global endpoint        
+        return Endpoint(
+            Url.parse("https://global.api"),
+        )
+    }
+}
+
+public class GetResourceResolveEndpointMiddleware(
+    private val config: SmithyService.Config,
+    private val request: GetResourceRequest,
+    private val provider: EndpointProvider,
+) : ModifyRequestMiddleware {
+    override suspend fun modifyRequest(req: SdkHttpRequest): SdkHttpRequest {
+        val params = SmithyServiceEndpointParams {
+            resourceId = request.resource
+        }
```

Comment:
**Comment**: Your above definition of `SmithyServiceEndpointParams` was a data class with a public constructor and no mention of a builder whereas here it's invoked using a builder lambda. The former is preferable if we can get away with it (e.g., backward compatibility isn't a concern).